File: .NET/Day 2/UserManagementAPI/UserManagementAPI/Models/ApiResponse.cs
using static UserManagementAPI.Enums;

namespace UserManagementAPI.Models
{
    public class ApiResponse<T>
    {
        public ResponseStatus ResponseStatus { get; set; }
        public string Message { get; set; } = string.Empty;
        public T? Result { get; set; }

        public static ApiResponse<T> Ok(string message, T? data = default)
            => new() { ResponseStatus = ResponseStatus.Success, Message = message, Result = data };

        public static ApiResponse<T> Fail(string message)
            => new() { ResponseStatus = ResponseStatus.Error, Message = message, Result = default };
    }
}

File: Node Js + Express Js/Day 2/express-app/src/shared/utils/api-response.ts
import { ApiResponse, ResponseStatus } from "../types/api-response.types.js";

export interface AppError extends Error {
  statusCode: number;
}

export const createAppError = (statusCode: number, message: string): AppError => {
  const error = new Error(message) as AppError;
  error.statusCode = statusCode;
  return error;
};

export const isAppError = (error: unknown): error is AppError => {
  return (
    error instanceof Error &&
    typeof (error as { statusCode?: unknown }).statusCode === "number"
  );
};

export const ok = <T>(message: string, result: T): ApiResponse<T> => {
  return {
    responseStatus: ResponseStatus.Success,
    message,
    result,
  };
};

export const fail = (message: string): ApiResponse<null> => {
  return {
    responseStatus: ResponseStatus.Error,
    message,
    result: null,
  };
};
